onnx tests failing:

```
torch/csrc/Module.cpp:521:3: error: 'END_HANDLE_TH_ERRORS' was not declared in this scope
11:08:51        END_HANDLE_TH_ERRORS
```

--- a/torch/csrc/DynamicTypes.cpp
+++ b/torch/csrc/DynamicTypes.cpp
@@ -2,7 +2,6 @@
 
 #include "DynamicTypes.h"
 #include "PythonTypes.h"
-#include "THP.h"
 #include "Exceptions.h"
 
 #include <vector>

--- a/torch/csrc/Generator.cpp
+++ b/torch/csrc/Generator.cpp
@@ -1,4 +1,5 @@
-#include <Python.h>
+#include "Generator.h"
+
 #include <structmember.h>
 #include <ATen/ATen.h>
 

--- a/torch/csrc/Generator.h
+++ b/torch/csrc/Generator.h
@@ -1,5 +1,4 @@
-#ifndef THP_GENERATOR_H
-#define THP_GENERATOR_H
+#pragma once
 
 #include <Python.h>
 #include <ATen/ATen.h>
@@ -29,6 +28,4 @@ THP_API PyObject *THPGeneratorClass;
 
 #ifdef _THP_CORE
 bool THPGenerator_init(PyObject *module);
-#endif
-
 #endif

--- a/torch/csrc/Module.cpp
+++ b/torch/csrc/Module.cpp
@@ -34,7 +34,7 @@
 #endif
 
 #define WITH_NUMPY_IMPORT_ARRAY
-#include "THP.h"
+#include "torch/csrc/utils/numpy_stub.h"
 
 namespace py = pybind11;
 
@@ -442,7 +442,7 @@ static PyObject* initModule() {
   ASSERT_TRUE(THPWrapper_init(module));
   ASSERT_TRUE(THPGenerator_init(module));
   ASSERT_TRUE(THPException_init(module));
-  ASSERT_TRUE(THPSize_init(module));
+  THPSize_init(module);
   ASSERT_TRUE(THPDtype_init(module));
   ASSERT_TRUE(THPVariable_initModule(module));
   ASSERT_TRUE(THPFunction_initModule(module));

--- a/torch/csrc/Module.cpp
+++ b/torch/csrc/Module.cpp
@@ -16,6 +16,7 @@
 #include <pybind11/pybind11.h>
 #include <pybind11/stl.h>
 
+#include "THP.h"
 #include "torch/csrc/DynamicTypes.h"
 #include "torch/csrc/DataLoader.h"
 #include "torch/csrc/Generator.h"

--- a/torch/csrc/Size.cpp
+++ b/torch/csrc/Size.cpp
@@ -1,11 +1,9 @@
 #include "Size.h"
 
 #include <string>
+#include "torch/csrc/utils/object_ptr.h"
 #include "torch/csrc/utils/python_strings.h"
 #include "torch/csrc/utils/python_tuples.h"
-#include "THP.h"
-
-PyObject* THPSizeClass = NULL;
 
 struct THPSize {
   PyTupleObject tuple;
@@ -13,13 +11,10 @@ struct THPSize {
 
 PyObject * THPSize_New(int dim, const int64_t *sizes)
 {
-  PyTypeObject* type = (PyTypeObject*)THPSizeClass;
-  PyObject* self = type->tp_alloc(type, dim);
-  if (!self) {
-    return NULL;
-  }
+  auto self = THPObjectPtr(THPSizeType.tp_alloc(&THPSizeType, dim));
+  if (!self) throw python_error();
   THPUtils_packInt64Array(self, dim, sizes);
-  return self;
+  return self.release();
 }
 
 static PyObject * THPSize_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
@@ -39,6 +34,7 @@ static PyObject * THPSize_pynew(PyTypeObject *type, PyObject *args, PyObject *kw
 
 static PyObject * THPSize_repr(THPSize *self)
 {
+  HANDLE_TH_ERRORS
   std::string repr("torch.Size([");
   for (Py_ssize_t i = 0; i < PyTuple_Size((PyObject*)self); ++i) {
     if (i != 0) {
@@ -48,6 +44,7 @@ static PyObject * THPSize_repr(THPSize *self)
   }
   repr += "])";
   return THPUtils_packString(repr);
+  END_HANDLE_TH_ERRORS
 }
 
 extern PyTypeObject THPSizeType;
@@ -138,12 +135,13 @@ PyTypeObject THPSizeType = {
   THPSize_pynew,                         /* tp_new */
 };
 
-bool THPSize_init(PyObject *module)
+void THPSize_init(PyObject *module)
 {
-  THPSizeClass = (PyObject*)&THPSizeType;
-  if (PyType_Ready(&THPSizeType) < 0)
-    return false;
+  if (PyType_Ready(&THPSizeType) < 0) {
+    throw python_error();
+  }
   Py_INCREF(&THPSizeType);
-  PyModule_AddObject(module, "Size", (PyObject *)&THPSizeType);
-  return true;
+  if (PyModule_AddObject(module, "Size", (PyObject*)&THPSizeType) < 0) {
+    throw python_error();
+  }
 }

--- a/torch/csrc/Size.h
+++ b/torch/csrc/Size.h
@@ -1,17 +1,14 @@
-#ifndef THP_SIZE_INC
-#define THP_SIZE_INC
+#pragma once
 
 #include <Python.h>
 #include "stdint.h"
 
-extern PyObject *THPSizeClass;
+extern PyTypeObject THPSizeType;
 
-#define THPSize_Check(obj) ((PyObject*)Py_TYPE(obj) == THPSizeClass)
+#define THPSize_Check(obj) (Py_TYPE(obj) == &THPSizeType)
 
 PyObject * THPSize_New(int dim, const int64_t *sizes);
 
 #ifdef _THP_CORE
-bool THPSize_init(PyObject *module);
-#endif
-
+void THPSize_init(PyObject *module);
 #endif

--- a/torch/csrc/Storage.cpp
+++ b/torch/csrc/Storage.cpp
@@ -12,6 +12,7 @@
 #include <TH/TH.h>
 #include <libshm.h>
 #include "THP.h"
+#include "allocators.h"
 #include "copy_utils.h"
 #include "DynamicTypes.h"
 

--- a/torch/csrc/THP.h
+++ b/torch/csrc/THP.h
@@ -36,7 +36,6 @@
 
 #ifdef _THP_CORE
 #include "serialization.h"
-#include "allocators.h"
 
 #include "autograd/autograd.h"
 #endif

--- a/torch/csrc/allocators.cpp
+++ b/torch/csrc/allocators.cpp
@@ -1,6 +1,6 @@
 #include <Python.h>
 
-#include "THP.h"
+#include "allocators.h"
 #include "torch/csrc/utils/auto_gil.h"
 
 // Adapted from fblualib
@@ -32,47 +32,6 @@ void StorageWeakRefAllocator::free(void* ptr) {
   delete this;
 }
 
-
-#ifdef WITH_NUMPY
-/**
- * Note [Numpy memory management]
- * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- * For efficiency reasons, when a user converts to/from numpy arrays,
- * we want to share the underlying storage.  This means that if we
- * turn a Numpy array into a Torch tensor, the Torch tensor must
- * keep the Numpy array alive, and vice versa for conversions in
- * the other direction.
- *
- * A Torch tensor keeps its backing Numpy array alive using the custom allocator
- * THNumpyArrayAllocator (backed by NumpyArrayAllocator), which holds a
- * THPObjectPointer to the Numpy PyArrayObject, and nulls it out upon free.
- * The relevant code is in torch/csrc/generic/Tensor.cpp.
- *
- * A Numpy array keeps its backing Torch tensor alive using the base object
- * <https://docs.scipy.org/doc/numpy-dev/reference/c-api.array.html#c.PyArray_SetBaseObject>
- * field of Numpy, which is Numpy's hook for allowing an external user to
- * manage memory.  The relevant code is in
- * torch/csrc/generic/methods/TensorSerialization.cwrap
- */
-
-// See Note [Numpy memory management]
-void* NumpyArrayAllocator::realloc(void* ptr, ptrdiff_t size) {
-  throw std::logic_error("NumpyArrayAllocator::realloc() not supported");
-}
-
-// See Note [Numpy memory management]
-void NumpyArrayAllocator::free(void* ptr) {
-  PyArrayObject *array_ptr = (PyArrayObject*)object.get();
-  if (!array_ptr || ptr != PyArray_DATA(array_ptr))
-    throw std::logic_error("invalid call to NumpyArrayAllocator::free()");
-  {
-    AutoGIL gil;
-    object = nullptr;
-  }
-  delete this;
-}
-#endif
-
 template<typename T>
 static void * malloc_wrapper(void *ctx, ptrdiff_t size) {
   return ((T*)ctx)->malloc(size);
@@ -99,15 +58,6 @@ THAllocator THStorageWeakRefAllocator = {
   realloc_wrapper<StorageWeakRefAllocator>,
   free_wrapper<StorageWeakRefAllocator>,
 };
-
-#ifdef WITH_NUMPY
-// See Note [Numpy memory management]
-THAllocator THNumpyArrayAllocator = {
-  malloc_wrapper<NumpyArrayAllocator>,
-  realloc_wrapper<NumpyArrayAllocator>,
-  free_wrapper<NumpyArrayAllocator>,
-};
-#endif
 
 #ifdef WITH_CUDA
 cudaError_t CudaStorageWeakRefAllocator::malloc(void** ptr, size_t size, cudaStream_t stream) {

--- a/torch/csrc/allocators.h
+++ b/torch/csrc/allocators.h
@@ -4,6 +4,7 @@
 #include <type_traits>
 #include <memory>
 
+#include <TH/TH.h>
 #ifdef WITH_CUDA
 #include <THC/THC.h>
 #endif

--- a/torch/csrc/allocators.h
+++ b/torch/csrc/allocators.h
@@ -1,6 +1,6 @@
-#ifndef THP_NUMPY_INC
-#define THP_NUMPY_INC
+#pragma once
 
+#include <Python.h>
 #include <type_traits>
 #include <memory>
 
@@ -8,16 +8,7 @@
 #include <THC/THC.h>
 #endif
 
-#ifdef WITH_NUMPY
-
-#ifndef WITH_NUMPY_IMPORT_ARRAY
-#define NO_IMPORT_ARRAY
-#endif
-#define PY_ARRAY_UNIQUE_SYMBOL __numpy_array_api
-#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
-#include <numpy/arrayobject.h>
-
-#endif
+#include "torch/csrc/utils/object_ptr.h"
 
 // Adapted from fblualib
 class ObjectPtrAllocator {
@@ -69,24 +60,8 @@ public:
 };
 #endif
 
-#ifdef WITH_NUMPY
-class NumpyArrayAllocator: public ObjectPtrAllocator {
-public:
-  NumpyArrayAllocator(PyObject *wrapped_array):
-      ObjectPtrAllocator(wrapped_array) {}
-
-  void* realloc(void* ptr, ptrdiff_t size);
-  void free(void* ptr);
-};
-#endif
-
 extern THAllocator THObjectPtrAllocator;
 extern THAllocator THStorageWeakRefAllocator;
 #ifdef WITH_CUDA
 extern THCDeviceAllocator THCStorageWeakRefAllocator;
-#endif
-#ifdef WITH_NUMPY
-extern THAllocator THNumpyArrayAllocator;
-#endif
-
 #endif

--- a/torch/csrc/autograd/functions/pybind.h
+++ b/torch/csrc/autograd/functions/pybind.h
@@ -4,9 +4,8 @@
 #include <pybind11/pybind11.h>
 #include <pybind11/stl.h>
 
+#include "torch/csrc/autograd/python_function.h"
 #include "torch/csrc/autograd/python_cpp_function.h"
-#include "torch/csrc/DynamicTypes.h"
-#include "torch/csrc/THP.h"
 
 namespace py = pybind11;
 
@@ -30,4 +29,3 @@ public:
 
 
 }} // namespace pybind11::detail
-

--- a/torch/csrc/autograd/init.cpp
+++ b/torch/csrc/autograd/init.cpp
@@ -1,28 +1,23 @@
 #include <Python.h>
+
+#include "torch/csrc/Exceptions.h"
 #include "torch/csrc/utils/pybind.h"
 #include "torch/csrc/autograd/grad_mode.h"
 #include "torch/csrc/autograd/profiler.h"
-
-#include "THP.h"
-
-#ifdef _MSC_VER
-#define ENSURE_UNREACHABLE __assume(0);
-#else
-#define ENSURE_UNREACHABLE __builtin_unreachable();
-#endif
+#include "torch/csrc/autograd/python_function.h"
 
 PyObject * THPAutograd_initExtension(PyObject *_unused)
 {
-  THPUtils_assert_PyImport("torch.autograd", autograd_module);
-  PyObject *autograd_dict = PyModule_GetDict(autograd_module);
+  auto autograd_module = THPObjectPtr(PyImport_ImportModule("torch.autograd"));
+  if (!autograd_module) throw python_error();
 
-  THPVariableClass      = PyMapping_GetItemString(autograd_dict,(char*)"Variable");
-  THPFunctionClass      = PyMapping_GetItemString(autograd_dict,(char*)"Function");
+  // NOTE: "leaks" Variable
+  THPVariableClass = PyObject_GetAttrString(autograd_module, "Variable");
+  if (!THPVariableClass) throw python_error();
 
-  THPUtils_assert(THPVariableClass, "couldn't find Variable class in "
-          "torch.autograd module");
-  THPUtils_assert(THPFunctionClass, "couldn't find Function class in "
-          "torch.autograd module");
+  // NOTE: "leaks" Function
+  THPFunctionClass = PyObject_GetAttrString(autograd_module, "Function");
+  if (!THPFunctionClass) throw python_error();
 
   auto m = py::handle(autograd_module).cast<py::module>();
 
@@ -62,7 +57,7 @@ namespace torch { namespace autograd {
 static PyObject * set_grad_enabled(PyObject* _unused, PyObject *arg) {
   HANDLE_TH_ERRORS
   if (!PyBool_Check(arg)) {
-    at::runtime_error("enabled must be a bool (got %s)", Py_TYPE(arg)->tp_name);
+    throw TypeError("enabled must be a bool (got %s)", Py_TYPE(arg)->tp_name);
   }
   GradMode::set_enabled(arg == Py_True);
   Py_RETURN_NONE;

--- a/torch/csrc/cuda/Storage.cpp
+++ b/torch/csrc/cuda/Storage.cpp
@@ -7,6 +7,7 @@
 #include "THCP.h"
 
 #include "override_macros.h"
+#include "torch/csrc/allocators.h"
 #include "torch/csrc/copy_utils.h"
 #include "DynamicTypes.h"
 

--- a/torch/csrc/utils.h
+++ b/torch/csrc/utils.h
@@ -124,16 +124,6 @@ THP_API void THPUtils_invalidArguments(
         PyObject *given_args, PyObject *given_kwargs,
         const char *function_name, size_t num_options, ...);
 
-#define THPUtils_assert_PyImport(name, module)                          \
-  PyObject* module = PyImport_ImportModule(name);                       \
-  if (!module) {                                                        \
-    if (PyErr_Occurred()) {                                             \
-      PyErr_Print();                                                    \
-    }                                                                   \
-    THPUtils_setError("class loader couldn't access %s", name);         \
-    return NULL;                                                        \
-  }
-
 #ifdef _THP_CORE
 
 bool THPUtils_checkIntTuple(PyObject *arg);

--- a/torch/csrc/utils/numpy_stub.h
+++ b/torch/csrc/utils/numpy_stub.h
@@ -1,0 +1,21 @@
+#pragma once
+
+#include <Python.h>
+
+#ifdef WITH_NUMPY
+
+#if !defined(NO_IMPORT_ARRAY) && !defined(WITH_NUMPY_IMPORT_ARRAY)
+#define NO_IMPORT_ARRAY
+#endif
+
+#ifndef PY_ARRAY_UNIQUE_SYMBOL
+#define PY_ARRAY_UNIQUE_SYMBOL __numpy_array_api
+#endif
+
+#ifndef NPY_NO_DEPRECATED_API
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#endif
+
+#include <numpy/arrayobject.h>
+
+#endif // WITH_NUMPY

--- a/torch/csrc/utils/python_arg_parser.h
+++ b/torch/csrc/utils/python_arg_parser.h
@@ -26,15 +26,16 @@
 #include <vector>
 #include <ATen/ATen.h>
 
-#include "torch/csrc/THP.h"
-#include "torch/csrc/utils/object_ptr.h"
-#include "torch/csrc/Exceptions.h"
-#include "torch/csrc/autograd/python_variable.h"
-#include "torch/csrc/utils/python_numbers.h"
 #include "torch/csrc/DynamicTypes.h"
 #include "torch/csrc/Dtype.h"
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/Generator.h"
+#include "torch/csrc/autograd/python_variable.h"
 #include "torch/csrc/autograd/generated/VariableType.h"
 #include "torch/csrc/tensor/python_tensor.h"
+#include "torch/csrc/utils/object_ptr.h"
+#include "torch/csrc/utils/python_numbers.h"
+#include "torch/csrc/utils/numpy_stub.h"
 
 namespace torch {
 
@@ -159,7 +160,8 @@ inline at::Tensor PythonArgs::tensor(int i) {
     // a test for Py_None here; instead, you need to mark the argument
     // as *allowing none*; you can do this by writing 'Tensor?' instead
     // of 'Tensor' in the ATen metadata.
-    throw TypeError("expected Variable as argument %d, but got %s", i, THPUtils_typename(args[i]));
+    throw TypeError("expected Variable as argument %d, but got %s", i,
+        Py_TYPE(args[i])->tp_name);
   }
   return reinterpret_cast<THPVariable*>(args[i])->cdata;
 }
@@ -191,7 +193,7 @@ inline std::vector<at::Tensor> PythonArgs::tensorlist(int i) {
     PyObject* obj = tuple ? PyTuple_GET_ITEM(arg, idx) : PyList_GET_ITEM(arg, idx);
     if (!THPVariable_Check(obj)) {
       throw TypeError("expected Variable as element %d in argument %d, but got %s",
-                 idx, i, THPUtils_typename(args[i]));
+                 idx, i, Py_TYPE(args[i])->tp_name);
     }
     res[idx] = reinterpret_cast<THPVariable*>(obj)->cdata;
   }
@@ -212,7 +214,7 @@ inline std::array<at::Tensor, N> PythonArgs::tensorlist_n(int i) {
     PyObject* obj = tuple ? PyTuple_GET_ITEM(arg, idx) : PyList_GET_ITEM(arg, idx);
     if (!THPVariable_Check(obj)) {
       throw TypeError("expected Variable as element %d in argument %d, but got %s",
-                 idx, i, THPUtils_typename(args[i]));
+                 idx, i, Py_TYPE(args[i])->tp_name);
     }
     res[idx] = reinterpret_cast<THPVariable*>(obj)->cdata;
   }

--- a/torch/csrc/utils/tensor_new.cpp
+++ b/torch/csrc/utils/tensor_new.cpp
@@ -5,15 +5,17 @@
 
 #include "torch/csrc/DynamicTypes.h"
 #include "torch/csrc/Exceptions.h"
+#include "torch/csrc/Size.h"
+#include "torch/csrc/autograd/variable.h"
 #include "torch/csrc/utils/auto_gil.h"
 #include "torch/csrc/utils/auto_gpu.h"
 #include "torch/csrc/utils/cuda_lazy_init.h"
+#include "torch/csrc/utils/numpy_stub.h"
 #include "torch/csrc/utils/python_arg_parser.h"
 #include "torch/csrc/utils/python_numbers.h"
 #include "torch/csrc/utils/python_scalars.h"
 #include "torch/csrc/utils/python_strings.h"
 #include "torch/csrc/utils/tensor_numpy.h"
-#include "torch/csrc/autograd/variable.h"
 
 static const int MAX_DIMS = 128;
 

--- a/torch/csrc/utils/tensor_numpy.cpp
+++ b/torch/csrc/utils/tensor_numpy.cpp
@@ -1,5 +1,7 @@
 #include "tensor_numpy.h"
 
+#include "torch/csrc/utils/numpy_stub.h"
+
 #ifndef WITH_NUMPY
 namespace torch { namespace utils {
 PyObject* tensor_to_numpy(const at::Tensor& tensor) {
@@ -19,11 +21,6 @@ at::Tensor tensor_from_numpy(PyObject* obj) {
 #include <memory>
 #include <sstream>
 #include <stdexcept>
-
-#define NO_IMPORT_ARRAY
-#define PY_ARRAY_UNIQUE_SYMBOL __numpy_array_api
-#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
-#include <numpy/arrayobject.h>
 
 using namespace at;
 using namespace torch::autograd;

